### DISCUSSION TITLE
Form widget

### DISCUSSION THREAD

**Discussion by lllama** [Textualize/textual#105] [General]
I'm trying to get my head around what a form widget would look like. For my little app, I'd like to have people enter some credentials, so I'd like a form with two rows and two columns. I'm struggling to work out how to put the pieces together.

My thinking so far:

* The whole thing gets wrapped in a `Panel`
* The entry fields would need to display underscores until they have some content
* The widget would need to support tabbing between the two fields and potentially other bindings
* Text entry would need to work somehow. I'm thinking just grabbing `on_key` and updating the content but would be curious if there was another way.
* The various pieces would need to be put in a grid: the `username` and `password` labels left-aligned in the left column and then the right column being left-aligned as well.

TBH, I'm not 100% sure how to put a mock-up together with Rich, so I'm going to start there but I'm posting here as it's some useful rubber-ducking if nothing else.

**Comment by lllama** [comment]
Looks like https://github.com/sirfuzzalot/textual-inputs might be something to draw inspiration from. (Spotted in an answer by Will in an issue somewhere.)

**Reply by aitoehigie** [reply]
@lllama did you finally make progress on this issue?
I am trying to implement a widget that takes file input and any pointers will be helpful

**Reply by lllama** [reply]
Sorry, no. Haven't looked at this in a while.

**Reply by davetapley** [reply]
⚠️ heads up to others: this is [pinned to very outdated textual](https://github.com/sirfuzzalot/textual-inputs/blob/9daf42f478c9a3820af82ca5ea7c2d308484b66a/pyproject.toml#L23) and so will likely break your code if you install it

**Comment by willmcgugan** [comment]
Form inputs are an important feature, but there's lots to do before I get there.

There probably needs to be a table layout, to have rows and columns of alternating labels. and widgets. Also a fleshed out tab focus system.

Ultimately I would like there to be a declarative approach, like Pydantic which would both run the widgets and returned validated data.

Lots to do first though...

**Reply by lllama** [reply]
Love the idea of declarative forms and pydantic or pydantic-like validation would be another win. Forms and form validation will always be a complete pain to do, so anything that can help with that is a bonus.
